# Analysis Plan, 12/3/23 check-in
#### Joseph Rush

### Explanatory text about the game, the mode, and the terminology will be inserted. I know that this draft currently has a lot of terms that are going to bounce off of readers, including you, Dr. Silva. The Analysis Plan content begins after the Weapon Data subheader

## The Data

#### Gunslinger & Sharpshooter Accuracy & Headshot Accuracy
Accuracy and headshot accuracy values for gunslinger & sharpshooter players, cleaned for data-distortions.

Additionally, the damage dealt by the player, the team's overall damage, and the % of the team damage dealt by that player (for use in weighting certain calculations down the line.)

Two-dimensional table, no nested layers. Gathered from the recent tournament as well as the spreadsheet maintained by a high-level, closed European play group.

In [1]:
import pandas as pd
dps_acc_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\GS&SS_acc.csv')
dps_acc_df.head()

ACC  HSACC (Hits)  HSACC (shots)      DD  Team DD   % of DD
0  68.790538     49.639490      34.147272  365402  1449721  0.252050
1  67.096135     37.444609      25.123885  300911  1449721  0.207565
2  60.382855     39.145417      23.637120  172028  1223218  0.140636
3  66.071429     61.261261      40.476190  535791  1440476  0.371954
4  74.509804     65.789474      49.019608  368733  1440476  0.255980

#### Commando Accuracy & Headshot Accuracy
Accuracy and headshot accuracy values for commando players.

Two-dimensional table, no nested layers. Gathered from the recent tournament as well as the spreadsheet maintained by a high-level, closed European play group.

In [2]:
mando_acc_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\mando_acc.csv')
mando_acc_df.head()

ACC  HSACC (hits)  HSACC (shots)      DD  Team DD   % of DD
0  57.986840     33.196523      19.249615  282885  1449721  0.195131
1  64.207650     37.624113      24.157559  267287  1223218  0.218511
2  51.885014     22.464426      11.655671  167241  1440476  0.116101
3  53.115819     20.436681      10.855111  181419  1424481  0.127358
4  46.297641     33.437868      15.480944  176153  1395982  0.126186

#### Zed Populations
For a given cycle, how many of each enemy type are spawned in a given wave.

Two-dimensional tables, no nested layers, gathered by analyzing the information given to the Controlled Difficulty Mod with the Tamari's SpawnCycler program.

In [3]:
bl2_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-bl_v2.csv')
nt2_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-nt_v2.csv')
osffi1_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-osffi_v1.csv')
apm2_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-apm_v2.csv')
mig2_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-ts_mig_v2.csv')
doom2pr_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-doom_v2_plus_rmk.csv')
mig3_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-ts_mig_v3.csv')
mb0a_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-mb_v0a.csv')
nvm1_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\populations-nvm_v1.csv')
nvm1_df.head()

Cyst  Alpha Clot  Slasher  Rioter  Gorefast  Gorefiend  Crawler  Stalker  \
0     0          31       26      12        31         18       39       39   
1     0          32       35      19        17         23       38       39   
2     0          38       30      21        21         26       36       44   
3     0          29       30      24        19         28       29       34   
4     0          30       28      27        15         32       28       37   

   Bloat  Husk  Siren  Scrake  Quarterpound  Fleshpound  
0     11    15     13       7             0           3  
1     14    20     14      16             0           7  
2     18    28     16      23             0          12  
3     19    29     20      34             0          18  
4     24    33     26      41             0          22

#### Zed Stats
Head health, headshot multiplier, & damage resistances (handgun, rifle, & assault rifle) for each zed type.

Two-dimensional tables, no nested layers, gathered from SimpleCat's KF2 Stats Spreadsheet

In [4]:
zedstats_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\ZedStats.csv')
zedstats_df.head()

Category  Cyst  Alpha Clot  Slasher  Rioter  Gorefast  \
0                   Head Health  20.0        20.0     20.0   125.0     50.00   
1           Headshot Multiplier   1.1         1.1      1.1     1.1      1.10   
2  AssaultRifle_AR15 resistance   1.5         1.0      1.0     1.0      1.20   
3      Assault Rifle resistance   1.5         1.0      1.0     1.0      1.20   
4              Rifle resistance   1.0         1.0      1.0     1.0      1.25   

   Gorefiend  Crawler  Stalker    Bloat     Husk  Siren  Scrake  \
0     150.00     20.0    20.00  82.0000  220.000  155.0  1584.0   
1       1.10      1.1     1.10   1.0001    1.001    1.1     1.1   
2       1.20      1.5     1.50   0.4000    0.650    1.0     1.0   
3       1.20      1.5     1.50   0.3500    0.650    1.0     1.0   
4       1.25      1.0     0.85   0.3500    0.700    1.0     1.0   

   Quarter Pound  Fleshpound  
0        1228.00     1716.00  
1           1.10        1.10  
2           0.62        0.50  
3           0.62        0.50  
4           0.75        0.75

#### Weapon Data
Cost for initial purchase, rounds received with initial purpose, cost per magazine, cost per bullet, base damage per bullet. Upgraded weapons are included as a line entry with no initial bullets that costs the price of the upgrade.

Two-dimensional tables, no nested layers, gathered from game data and SimpleCat's KF2 Stats Spreadsheet

In [5]:
weapon_data_df = pd.read_csv('E:\College\EMAT32110_2-DataInEmergingMediaAndTechnology\FinalProjectData\WeaponStats.csv')
weapon_data_df.head()

Weapon  Purchase cost  Purchase ammo  Cost per magazine  \
0        AR15            200            120               20.0   
1     Bullpup            650             90               30.0   
2        SCAR           1500             80               28.0   
3      FN FAL           1500             60               47.0   
4  Winchester            200             48               32.0   

   Bullets per magazine  Max Ammo (mag+spare)  Cost per bullet  Base damage  \
0                    20                   308           1.0000           30   
1                    32                   354           0.9375           32   
2                    20                   428           1.4000           55   
3                    20                   428           2.3500           70   
4                    12                    96           2.6667           80   

  Damage Type  
0     Assault  
1     Assault  
2     Assault  
3       Rifle  
4       Rifle

## Assumptions
- Standard precision team comp: 1 commando, 2 gunslingers, 2 sharpshooters, as well as 1 medic
- Meta weapons
 - 9mm, Dual 1858, SPX(+2), dual deagles, and single AF for gunslinger
 - 9mm, Winchester, EBR(+1), SPX(+2) for sharpshooter
 - 9mm, AR15, Bullpup, SCAR, FN FAL for commando
- Unmodded weapons
- Weapon damage modeled at a consistent 5 REU. This isn’t quite accurate to real play, particularly for players with lower accuracy or playing parts of the map where they don’t have a constant flow of enemies to shoot, but it’s close enough to reality for this model. A more thorough model would look at shots fired per wave and real wave durations and perhaps model REU based on how many 2 second intervals go by without an headshots expected to be hit 
- Sharpshooters use 4 freeze grenades each in every wave. This is very realistic to actual play. If anything, they tend to use fewer. 
- All players only fill SPX to 80 rounds
- All zeds are killed by players
- Medics: see Tradeoff #1

#### Commando Damage Skills
- Class base
 - +25% perk weapon damage
- RLLLR until SCAR. 
 - Fallback: Increase damage with your 9mm pistol/HRG 93R and knife 85%; increase weapon switching speed 50%.
 - Hollow Point Rounds: Increase damage of perk weapons 30%
- RLLRR after SCAR
 - Fallback: Increase damage with your 9mm pistol/HRG 93R and knife 85%, and increase weapon switching speed 50%.
 
#### Gunslinger Damage Skills
- Class base
 - +25% perk weapon damage
- RLLRR skills
 - Rack ‘em Up: Consecutive headshots with perk weapons increases damage 10%, to a maximum of 50%
 
#### Sharpshooter Damage Skills
- Class base
 - +25% headshot damage
- RLLRR skills
 - Stability: +30% damage
 - Rack ‘em Up: Consecutive headshots with perk weapons increases damage 10%, to a maximum of 50%

## Tradeoffs:
1.	Headshot damage only. It is often reasonable to kill the weak enemies in-game with body shots. Hit zone data does exist for zed kills, but it does not account for prior shots or, in the case of larges, for decapitations before the body was killed. A fleshpound cannot be assumed to have been killed with body shots when it was actually decapitated first and then shot in the chest for the quadruple decapped body multiplier. Additionally, precision play emphasizes headshots, as they are most effective, and for the sake of not expanding the model into intractably unknowable territory, only headshots will be applied to zeds.
2.	Medics excerpted, and they are an extremely valuable part of team eco. They will be assumed for any amount of dosh required for both sharps to load up on w2 EBR and freezes. To include medics, I would need real world wave by wave eco data for them, and I don’t have that. The medic is then assumed to be able to contribute that much dosh to the team’s pool each wave.
3.	Assumes that overkill (wasted) damage and penetration damage roughly cancel each other out. This is not necessarily true in practice, but there is no good way to track them available either in-game or in the accumulated data.
4.	Weighting for zed damage resistances would be best derived from actual wave by wave kills per perk per cycle, but that’s beyond the scope of the easily accessible data.
5.	Only roughly accounts for target prioritization
6.	Currently does not account for shared money from kills. I might distribute 40% of large kill dosh across the trinity (mando, sharp 1, slinger 1) to account for some of this, roughly modeling Nausea
7.	Grenade damage for mando & gunslinger not included, since it is incredibly variable
8.	Dead-Eye ignored for sharpshooter. Its impact is unreliable, since it depends on individual playstyles, maps, and situations, so its existence is being disregarded.
9.	Cost per bullet isn’t quite exact to in-game modeling, since I used cost per magazine and there’s a little more stuff happening under the hood than that, but it’s close enough for this.
10.	The assumption that all zeds are killed by players is inaccurate to Friendly Fire, particularly in the case of Fleshpounds dealing damage to other zeds while raging or raged. However, zeds also have a chance to rout around the fleshpounds or move out of their way, so the modeling for this is intractable without massive amounts of game data of per wave zed kills by players on any particular cycle. The difference between the population and the players’ kills would give the number of zeds killed by friendly fire.
11.	Only one build order per perk:

##### Slinger
1. 1858s & 9mm
2. SPX, 1858s (rebought), & 9mm
3. Dual deags, SPX, 1858s (rebought), & 9mm
4. Dual deags, SPX+2, & 9mm
5. Add spare deag
6. Add single AF
7. Add spare deag
8. Add spare SPX+2

##### Sharp
1. Winchester & 9mm
2. EBR, Winchester, & 9mm; 6 freezes
3. EBR, EBR, Winchester, & 9mm; 6 freezes
4. EBR+1, EBR+1, Winchester, & 9mm; 6 freezes
5. SPX+2, EBR+1, EBR+1, 9mm; 6 freezes
6. Add EBR+1
7. Add SPX+2
8. Add EBR+1

##### Mando
1. AR15 & 9mm
2. Bullpup, AR15, & 9mm
3. SCAR, Bullpup, & 9mm
4. SCAR, FN FAL, & 9mm
5. Add Spare SCAR
6. Add Spare FAL
7. Add 9mm upgrade(s)
8. Add spare FAL

(I expect the model to hit the NVM_v1 w10 cutoff around step 6 or 7)
## Modeling
Presentation of damage formula
### Data gathering
1.	Gather acc & hsacc from summer tournament & EU NCZ group. All sharps & slingers in one group, all mandos in another. Trim out Tamariland significant AF users due to data distortion. (Significant defined as them buying dual AFs)
2.	Gather final zed kill totals per player from tournament data. Separate into two pools: sharp/slinger & mando
3.	Final damage per player. Per wave, if available.
4.	Zed population data for bl_v2, nt_v2, osffi_v1, apm_v2, ts_mig_v2, doom_v2_plus_rmk, ts_mig_v3, nvm_v1
5.	Weapon stats 
 - Cost per weapon
 - Cost per bullet
 - Base damage per weapon
 - Cost per upgrade
 - Damage increase per weapon
 - Upgraded base damage per weapon per upgrade

### Data Analysis
1.	__**SCATTER PLOT**__ of acc vs. hsacc
2.	Group acc/hsacc data with analysis to make a few representative player skill levels.
3.	Find _Weighted Damage Resistance Values_ i.e. FPs are more resistant to handgun than rifle. Use percentage of a given zed in the wave averaged with percentage of overall game kills from tourney data for the weight of that zed’s resistance value. If time, use tourney data for kills per wave.
4.	Find the adjusted zed head health pool for a given wave. [Head health]/[headshot multiplier] for every zed. This results in a big number.
5.	Find also: adjusted zed head health pool for NVM_v1 w10. This acts as the cutoff for the eco model.
6.	Find damage per bullet for each weapon before zed resistances are accounted for with standard skills at 5 racks where REU is relevant
7.	Multiply by current ammunition size to get _Max Damage Per Weapon_
8.	Find the Max Damage per Player for a given wave. This is [damage for every available bullet in a weapon]*[weighted resistance value] + the same for all their other weapons.
9.	Add together all the _Max Damage per Player_ values to get _Max Damage for Team_.
10.	Use the Max Damage Per Weapon derive an Expected Damage Weight percentage.
11.	Multiply the above value by hsacc% of the sample skill level for each player to get Expected Damage Per Player.
12.	Add together the expected damage of each player to get _Expected Team Damage Pool_.
13.	Subtract _Adjusted Zed Head Health Pool_ and _Expected Team Damage_.
14.	__**BAR PLOT**__ of _Max Damage for Team_ vs. _Expected Team Damage_ vs. _Adjusted Zed Head Health Pool_.
15.	Use the damage per perk data, either per wave or final in the matches, to create a _Damage Pool Weight_ that represents each perk’s expected contribution to the total damage. This helps account for target priority, where the mando in particular is firing at many low health targets and largely ignoring high health targets.
16.	Divide the spent damage among the players using the _Damage Pool Weight__ and their weapons to get _Damage Per Player-_ and spread across their weapons to get _Expected Ammunition Expenditure_. (This will weight 9mm higher than actual play. That is an acceptable tradeoff)
17.	If a wave total wants too much damage from a player, such as a mando not having enough ammo to deal the expected damage, distribute the overflow among the other players equally.
18.	Use the _Damage per Player_ and their HSACC% to find their ammo expenditure
19.	Find the _Minimum Dosh Total_, which is the minimum amount of money the team can earn if every zed is killed with no teamwork.
20.	Divide the _Minimum Dosh Total_ among the players. Buy, fill, sell, & upgrade weapons from the model build orders.
21.	Use any leftover money from players to buy ammo, freeze grenades, & upgrades for other players (this will have to be by hand. Writing an optimization algorithm to maximize team damage output is beyond my current skills).
22.	Use any leftover money to buy grenades for non-sharpshooters
23.	Save any remaining money for future use
24.	Repeat with next wave
25.	_**LINE PLOT**_ of team money per wave.
26.	Conglomerate _**LINE PLOT**_ of team money per wave for different cycles.
27.	When _Expected Team Damage Pool_ is greater than _NVM_v1 w10 Adjusted Head Health Pool_, cut the model off. Players, if they play well, can deal with basically any number of zeds in waves after this point. It is the "Git Gud" Threshold.